kythe/rust/extractor/src/lib.rs

```diff
@@ -29,75 +36,81 @@ use std::path::PathBuf;
+    let ofile = matches.opt_str("o").map(|o| PathBuf::from(&o));
 
-/// Handles compiler callbacks to enable and dump the save_analysis
-#[derive(Default)]
-struct CallbackShim {
-    output_dir: PathBuf,
-}
+    // Process configuration flags
+    let cfg = interface::parse_cfgspecs(matches.opt_strs("cfg"));
 
-impl CallbackShim {
-    /// Create a new CallbackShim that dumps save_analysis files to `output_dir`
-    pub fn new(output_dir: PathBuf) -> Self {
-        Self { output_dir }
-    }
-}
+    // Create PathBufs for the input file
+    let ifile = &matches.free[0];
 
-impl Callbacks for CallbackShim {
-    // Always enable save_analysis generation
-    fn config(&mut self, config: &mut interface::Config) {
-        config.opts.debugging_opts.save_analysis = true;
-    }
+    // Create the compilation configuration
+    let config = interface::Config {
+        opts: sopts,
+        crate_cfg: cfg,
+        input: Input::File(PathBuf::from(ifile)),
+        input_path: Some(PathBuf::from(ifile)),
+        output_file: ofile,
+        output_dir: odir,
+        file_loader: None,
+        diagnostic_output: DiagnosticOutput::Default,
+        stderr: None,
+        lint_caps: Default::default(),
+        parse_sess_created: None,
+        register_lints: None,
+        override_queries: None,
+        make_codegen_backend: None,
+        registry: Registry::new(&rustc_error_codes::DIAGNOSTICS),
+    };
 
-    fn after_analysis<'tcx>(
-        &mut self,
-        compiler: &interface::Compiler,
-        queries: &'tcx Queries<'tcx>,
-    ) -> Compilation {
-        let input = compiler.input();
-        let crate_name = queries.crate_name().unwrap().peek().clone();
+    interface::run_compiler(config, |compiler| {
+        compiler.enter(|queries| -> Result<()> {
+            queries.parse().map_err(|_| anyhow!("Failed to parse code"))?;
+            queries.expansion().map_err(|_| anyhow!("Failed to expand code"))?;
+            queries.prepare_outputs().map_err(|_| anyhow!("Failed to prepate outputs"))?;
+            queries.global_ctxt().map_err(|_| anyhow!("Failed to generate global context"))?;
 
-        // Configure the save_analysis to include full documentation.
-        // Normally this would be set using a `rls_data::config::Config` struct on the
-        // fourth parameter of `process_crate`. However, the Rust compiler
-        // falsely claims that there is a mismatch between rustc_save_analysis's
-        // `rls_data::config::Config` and ours, even though we use the same version.
-        // This forces us to use the environment variable method of configuration
-        // instead.
-        std::env::set_var(
-            "RUST_SAVE_ANALYSIS_CONFIG",
-            r#"{"output_file":null,"full_docs":true,"pub_only":false,"reachable_only":false,"distro_crate":false,"signatures":false,"borrow_data":false}"#,
-        );
+            std::env::set_var(
+                "RUST_SAVE_ANALYSIS_CONFIG",
```

Comment:
I'm not fully following the flow here, but this doesn't need to happen *before* `run_compiler` is invoked?

Reply:
This environment variable isn't read until we actually process the crate with the save analysis crate, so we set it right before we do the analysis.